power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp

```diff
@@ -0,0 +1,471 @@
+            msgpack::object const& obj = msg_data[element_number_];
+            if (obj.type == msgpack::type::ARRAY) {
+                parse_array_element(element_pointer, obj, attributes);
+            } else if (obj.type == msgpack::type::MAP) {
+                parse_map_element(element_pointer, obj, component);
+            } else {
+                throw SerializationError{"An element can only be a list or dictionary!\n"};
+            }
+        }
+        element_number_ = -1;
+    }
+
+    void parse_array_element(void* element_pointer, msgpack::object const& obj,
+                             std::span<MetaAttribute const* const> attributes) const {
+        auto const arr = obj.as<ArraySpan>();
+        if (arr.size() != attributes.size()) {
+            throw SerializationError{
+                "An element of a list should have same length as the list of predefined attributes!\n"};
+        }
+        for (attribute_number_ = 0; attribute_number_ != static_cast<Idx>(attributes.size()); ++attribute_number_) {
+            parse_attribute(element_pointer, arr[attribute_number_], *attributes[attribute_number_]);
+        }
+        attribute_number_ = -1;
+    }
+
+    void parse_map_element(void* element_pointer, msgpack::object const& obj, MetaComponent const& component) const {
+        for (msgpack::object_kv const& kv : obj.as<MapSpan>()) {
+            attribute_key_ = key_to_string(kv);
+            Idx const found_idx = component.find_attribute(attribute_key_);
+            if (found_idx < 0) {
+                attribute_key_ = "";
+                continue; // allow unknown key for additional user info
+            }
+            parse_attribute(element_pointer, kv.val, component.attributes[found_idx]);
+        }
+        attribute_key_ = "";
+    }
+
+    void parse_attribute(void* element_pointer, msgpack::object const& obj, MetaAttribute const& attribute) const {
+        // skip for none
+        if (obj.is_nil()) {
+            return;
+        }
+        // call relevant parser
+        switch (attribute.ctype) {
+        case CType::c_double:
+            return parse_attribute_per_type<double>(element_pointer, obj, attribute);
+        case CType::c_double3:
+            return parse_attribute_per_type<RealValue<false>>(element_pointer, obj, attribute);
+        case CType::c_int8:
+            return parse_attribute_per_type<int8_t>(element_pointer, obj, attribute);
+        case CType::c_int32:
+            return parse_attribute_per_type<int32_t>(element_pointer, obj, attribute);
+        default:
+            throw SerializationError{"Unknown data type for attriute!\n"};
+        }
+    }
+
+    template <class T>
+    void parse_attribute_per_type(void* element_pointer, msgpack::object const& obj,
+                                  MetaAttribute const& attribute) const {
+        obj >> *reinterpret_cast<T*>(reinterpret_cast<char*>(element_pointer) + attribute.offset);
```

Comment:
if we re-write this as `*reinterpret_cast<T*>(reinterpret_cast<char*>(element_pointer) + attribute.offset) = ...` (either by creating a wrapper function or by implementing the `operator>>` in the `MetaAttribute`), can we get rid of the double `reinterpret_cast`?

Reply:
I have added this in the attribute meta data.

The whold file is about using `msgpack`. I would keep using `operator>>` here as recommended in `msgpack`.